src/DotNetLightning.Core/Utils/Primitives.fs

```diff
@@ -117,30 +120,43 @@ module Primitives =
             ec |> LNECDSASignature
 
 
-    type PaymentPreimage = | PaymentPreimage of byte [] with
-        member x.Value = let (PaymentPreimage v) = x in v
+    type PaymentHash = | PaymentHash of uint256 with
+        member x.Value = let (PaymentHash v) = x in v
+        member x.ToBytes(?lEndian) =
+            let e = defaultArg lEndian true
+            x.Value.ToBytes(e)
+
+        member x.GetRIPEMD160() =
+            let b = x.Value.ToBytes()
+            Crypto.Hashes.RIPEMD160(b, b.Length)
+
+
+    type PaymentPreimage(value: seq<byte>) =
+        do
+            if value.Count() <> PaymentPreimage.LENGTH then
+                raise <| ArgumentException(sprintf "Payment preimage length should be %i" PaymentPreimage.LENGTH)
```

Comment:
Validating the length here is not bad. in fact it is better than before that I did not have any validation.
But returning result is better. such as
```fsharp
type PaymentPreimage = private PaymentPreimage of seq[]

module PaymentPreimage =
  let (|PaymentPreimage|_) x =
    match x with
    | PaymentPreimage x -> Some x
    | _ -> None

  let TryCreate (data: seq<bytes>) =
    // validate and return result
```

Reply:
I'm not familiar with this `let (|foo|) bar =` thing, what is that?

Reply:
Its active pattern (may be it was `let (|foo|_|) bar =` ) it is necessary for pattern matching in the case of private constructor.

Reply:
can you point me to any docs about this?

Reply:
https://habr.com/en/post/458730/

Reply:
Although the example explained in this post is for non-single DU. it has an use case in single case DU when you deconstruct PaymentPreimage in function argument. e.g.

```fsharp
let (PaymentPreimage x) =
    // do something with x
```

Speaking precisely you don't need module in this case. just place active pattern under the type definition. And TryCreate can be static method instead.

Reply:
it seems the types created with `type Foo = private Foo of Bar<Baz>` which later can have a `with` clause following augmentations, don't allow `let` declarations inside. Or maybe I'm missing something? So we may need a module after all.

Reply:
What I meant is,  in this case `type PaymentPreimage` is already placed in another module, so we can define active pattern directly *next* to the PaymentPreimage itself. (otherwise if PaymentPreiamge is placed under namespace, we have to prepare another module since it is impossible to place active pattern directly under namespace.)

Reply:
I'm sorry but I still don't get what you mean. Anyway, the current master branch doesn't have an active pattern on this type, so it's not being used at the moment. Either we should only add it when it's needed, or it can be added outside of this PR.

Reply:
Argh now the tests don't compile, is it because I need the active pattern?